app/src/main/java/me/hufman/androidautoidrive/carapp/music/TextScroller.kt

```diff
@@ -0,0 +1,52 @@
+package me.hufman.androidautoidrive.carapp.music
+
+/**
+ * Wrapper around a string that handles scrolling when it is too long to fit on a line.
+ */
+class TextScroller(private val originalText: String, private val maxLineLength: Int) {
+	companion object {
+		const val SCROLL_COOLDOWN_SECONDS = 6
+		const val INDEX_JUMP_VALUE = 3
+	}
+
+	val shouldScroll: Boolean
+	var scrollText: Boolean = false
+	var startIndex: Int = 0
+	var previousTimestamp: Long
+
+	init {
+		previousTimestamp = System.currentTimeMillis()
+		shouldScroll = originalText.length > maxLineLength
+	}
+
+	/**
+	 * Retrieves the current slice of the text to display. Every [SCROLL_COOLDOWN_SECONDS] the text
+	 * will be ready to start scrolling and making a call to this method will get the current state
+	 * of the text and shift the displayed slice at rate of [INDEX_JUMP_VALUE]. Once the slice reaches
+	 * the end of the original text, it will reset the slice back to the beginning and begin the
+	 * cooldown interval.
+	 *
+	 * If the original text is not long enough to need scrolling the original will be returned.
+	 */
+	fun getText(): String {
+		if (!shouldScroll) {
+			return originalText
+		}
+		var displayText = originalText
+		val endIndex = maxLineLength + startIndex
+		if (scrollText) {
+			if (endIndex <= originalText.length) {
+				displayText = originalText.substring(startIndex, endIndex)
```

Comment:
Does it look good with the ellipses character added to the start/end as necessary?

Reply:
I think only ellipses the start, and rely on the car's default clipping/fade effect for length... Can TextScroller be used in a mode that doesn't enforce a max length, for the fields that work fine with long text?

Reply:
The proposed uses of TextScroller are all in car widgets where the car will fade the end of the text instead of wrapping, and so we don't need to hard clip the text ourselves. Let's use the second substring call for both branches of the conditional: `originalText.substring(startIndex, originalText.length)`

Reply:
Sorry, bit of time away from my computer but I'm back. But that's a good idea to not hard clip the text and instead use the fading effect that is already offered, updated this and tested it out and it looks even better with that fade effect!

As far as the ellipses character added for long strings I haven't explicitly tested that scenario out but if we are to use the TextScroller for that I'd imagine we'd run into some performance issues since those ellipsis are usually added for items in a `RHMIComponent.List`. I think we might need to tailor a better solution for that scenario so I left it out of this initial design while I do some further investigation and testing there